saleor/checkout/complete_checkout.py

```diff
@@ -1213,3 +1222,140 @@ def create_order_from_checkout(
                 checkout_info.voucher, checkout_info.checkout.get_customer_email()
             )
             raise
+
+
+def complete_checkout(
+    manager: "PluginsManager",
+    checkout_info: "CheckoutInfo",
+    lines: Iterable["CheckoutLineInfo"],
+    payment_data,
+    store_source,
+    discounts,
+    user,
+    app,
+    site_settings=None,
+    tracking_code=None,
+    redirect_url=None,
+    metadata_list: Optional[List] = None,
+    private_metadata_list: Optional[List] = None,
+) -> Tuple[Optional[Order], bool, dict]:
+    """Logic required to finalize the checkout and convert it to order.
+
+    Should be used with transaction_with_commit_on_errors, as there is a possibility
+    for thread race.
+    :raises ValidationError
+    """
+    with transaction_with_commit_on_errors():
+        payment, customer_id, order_data = complete_checkout_pre_payment_part(
+            manager=manager,
+            checkout_info=checkout_info,
+            lines=lines,
+            discounts=discounts,
+            user=user,
+            site_settings=site_settings,
+            tracking_code=tracking_code,
+            redirect_url=redirect_url,
+        )
+        _reserve_stocks_without_availability_check(checkout_info, lines)
+
+    # Process payments out of transaction to unlock stock rows for another user,
+    # who potentially can order the same product variants.
+    txn = None
+    if payment:
+        txn = _process_payment(
+            payment=payment,
+            customer_id=customer_id,
+            store_source=store_source,
+            payment_data=payment_data,
+            order_data=order_data,
+            manager=manager,
+            channel_slug=checkout_info.channel.slug,
+        )
+
+    with transaction_with_commit_on_errors():
+        # Run pre-payment checks to make sure, that nothing has changed to the
+        # checkout, during processing payment.
+
+        checkout_info.checkout.voucher_code = None
+        _, _, post_payment_order_data = complete_checkout_pre_payment_part(
+            manager=manager,
+            checkout_info=checkout_info,
+            lines=lines,
+            discounts=discounts,
+            user=user,
+            site_settings=site_settings,
+            tracking_code=tracking_code,
+            redirect_url=redirect_url,
+        )
+        if _compare_order_data(order_data, post_payment_order_data):
+            order, action_required, action_data = complete_checkout_post_payment_part(
+                manager=manager,
+                checkout_info=checkout_info,
+                lines=lines,
+                payment=payment,
+                txn=txn,
+                order_data=order_data,
+                user=user,
+                app=app,
+                site_settings=site_settings,
+                metadata_list=metadata_list,
+                private_metadata_list=private_metadata_list,
+            )
+        else:
+            release_voucher_usage(
```

Comment:
Question: If we are releasing a voucher and payment, in case of mismatch between order_data. Shouldn't we also drop the reservation for the checkout?

Reply:
We should drop it if reservations are disabled.

Reply:
Reservation is done for a short time, currently set on 45 seconds. As I recall, it is a max time for processing payments.